modules/internal/ChapelDistribution.chpl

```diff
@@ -119,47 +159,113 @@ module ChapelDistribution {
     // The common case seems to be local access to this class, so we
     // will use explicit processor atomics, even when network
     // atomics are available
-    var _domCnt: atomic_refcnt; // domain reference count
     var _arrs: list(BaseArr);  // arrays declared over this domain
+    var _arrs_containing_dom: int; // number of arrays using this domain
+                                   // as var A: [D] [1..2] real
+                                   // is using {1..2}
     var _arrsLock: atomicbool; //   and lock for concurrent access
+    var _free_when_no_arrs: bool;
+    var pid:int = -1; // privatized ID, if privitization is supported
   
+    proc ~BaseDom() {
+    }
+
     proc dsiMyDist(): BaseDist {
       halt("internal error: dsiMyDist is not implemented");
       return nil;
     }
   
+    // Returns (dom, dist).
+    // if this domain should be deleted, dom=this; otherwise it is nil.
+    // dist is nil or a distribution that should be removed.
     pragma "dont disable remote value forwarding"
-    proc destroyDom(): int {
-      compilerAssert(!noRefCount);
-      var cnt = decRefCount();
-      if cnt == 0 && dsiLinksDistribution() {
-          var dist = dsiMyDist();
-          on dist {
-            local dist.remove_dom(this);
-            var cnt = dist.destroyDist();
-            if cnt == 0 then
-              delete dist;
+    proc remove() : (BaseDom, BaseDist) {
+
+      // TODO -- remove dsiLinksDistribution
+      assert( dsiMyDist().dsiTrackDomains() == dsiLinksDistribution() );
```

Comment:
yes. I don't want to do it right now since it's not strictly required.
My thought is that it makes more sense to have dist.dsiTrackDomains() than domain.dsiLinksDistribution() but others might disagree. For now we have both.